internal/rbd/mirror.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+)
+
+// enableImageMirroring enables mirroring on an image.
+func (rv *rbdVolume) enableImageMirroring(ctx context.Context, mode librbd.ImageMirrorMode) error {
+	util.DebugLog(ctx, "rbd: enable mirroring %s with mode %d", rv.String(), mode)
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+
+	err = image.MirrorEnable(mode)
+	return err
+}
+
+// disableImageMirroring disables mirroring on an image.
+func (rv *rbdVolume) disableImageMirroring(ctx context.Context, force bool) error {
```

Comment:
There has been some movement to reduce the passing of a context to rbdVolume methods. It is better to return useful errors with details and have callers log those. Passing a context to internal rbdVolume methods makes it sometimes more difficult to re-use the functions.

Reply:
Agree, we are already logging the error message if something fails. but without logging with context makes it hard for debugging if we call multiple API if one is successful and one is stuck it's very difficult to debug it.

Reply:
Indeed, that is why returning useful and complete error messages is important. Not everyone has debugging enabled, and also in those environments the issue needs to be identified. Callers will likely include the action that failed (disable image mirroring), but this function should probably mention if `open()` or `MirrorDisable()` failed.

We're not doing that consistently everywhere, but maybe we can improve the new code we're adding.

Reply:
removed ctx in mirror.go file